components/script/dom/bindings/structuredclone.rs

```diff
@@ -129,12 +132,15 @@ unsafe extern "C" fn read_callback(cx: *mut JSContext,
                                    r: *mut JSStructuredCloneReader,
                                    tag: u32,
                                    _data: u32,
-                                   _closure: *mut raw::c_void)
+                                   closure: *mut raw::c_void)
                                    -> *mut JSObject {
     assert!(tag < StructuredCloneTags::Max as u32, "tag should be lower than StructuredCloneTags::Max");
     assert!(tag > StructuredCloneTags::Min as u32, "tag should be higher than StructuredCloneTags::Min");
+    let holder = Rc::from_raw(closure as *const StructuredCloneHolder);
     if tag == StructuredCloneTags::DomBlob as u32 {
-        return read_blob(cx, r)
+        let (rooted_blob, js_object) = read_blob(cx, r);
+        holder.blobs.borrow_mut().push(rooted_blob);
```

Comment:
We need to be very careful here, because if `holder.blob` is being borrowed mutably or immutably elsewhere, this will panic. Specifically, there may be a chance that some code elsewhere borrows it immutably for reading, and then during the read, it performs yet another StructuredCloneRead, which would then run into a panic.

Reply:
Should we use `try_borrow_mut` and do some error handling at this point?

Reply:
That sounds good to me.